### PR TITLE
ServicesManager should use ServiceRegistry for find operations

### PR DESCRIPTION
**Pull request by DimaAy**
This pull request fixes the fact that find operations of the ServiceRegistry is not fully utilized by the ServicesManager instead it does its own filtering logic.

### CODE REVIEW

**Comment by mmoayyed** [comment]
Thanks very much.

This is close, but not all the way. The main issue is that internal services collection owned by the manager is disconnected from the registry calls. 

Consider the following scenario: 

- The manager's scheduler is disabled to periodically load services from the registry.
- On startup, CAS calls the manager to load the services from the registry once.
- All find() operations of the manager will now use the internal list of services, never calling the registry (OK)
- Someone adds a new service definition S to the source.
- The manager attempts to find S, failing to find it in the internal services list, calls the registry. Registry returns S.
- S is never put into the internal list. All calls for S from the manager are routed to the registry
- CAS dies.

Or consider the following scenario:


- The manager's scheduler is disabled to periodically load services from the registry.
- On startup, CAS calls the manager to load the services from the registry once.
- All find() operations of the manager will now use the internal list of services, never calling the registry (OK)
- Someone adds a new service definition S to the source.
- The manager attempts to find S, failing to find it in the internal services list, calls the registry. Registry returns S.
- S is added to the internal list of services. (New change)
- Someone updates S in the source.
- Changes to S are never picked by CAS, because the entry is always cached and returned in the internal list.
- CAS dies.

What you should do is,

- When the registry finds a service, you should add/update it in the list. However, this only fixes the first scenario. Instead, you should turn the list/map of services into a cache backed by Caffeine, which would help in scenarios. The cache will auto-expire entries at some point, allowing items to be refreshed, or in general allows the cache to be 100% disabled for real-time updates.

**Comment by DimaAy** [comment]
Thanks for the review, I got your points, I am going to take care of those two scenarios.

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/services/ServiceRegistryProperties.java [hunk 1]
```diff
@@ -144,7 +144,22 @@
...
+    /**
+     * services cache size.
+    */
+    private long cacheSize = Long.MAX_VALUE;
 
+    /**
+     * services cache capacity.
+    */
+    private int cachCapacity = Integer.MAX_VALUE;
+
```
Comment (mmoayyed) [inline api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/services/ServiceRegistryProperties.java:162]: Seems like you should explain these in much more detail.

File: api/cas-server-core-api-services/src/main/java/org/apereo/cas/services/ServicesManager.java [hunk 2]
```diff
@@ -132,6 +132,38 @@ default void save(final RegisteredService... services) {
         }
         return null;
     }
+    
+    /**
+     * Find a RegisteredService by exact service id.
```
Comment (mmoayyed) [inline api/cas-server-core-api-services/src/main/java/org/apereo/cas/services/ServicesManager.java:137]:
```suggestion
     * Find a {@link RegisteredService} by exact service id.
```

File: api/cas-server-core-api-services/src/main/java/org/apereo/cas/services/ServicesManager.java [hunk 3]
```diff
@@ -132,6 +132,38 @@ default void save(final RegisteredService... services) {
...
+     *
+     * @param serviceId the service
+     * @return the RegisteredService
```
Comment (mmoayyed) [inline api/cas-server-core-api-services/src/main/java/org/apereo/cas/services/ServicesManager.java:140]:
```suggestion
     * @return the service definition or null
```

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 4]
```diff
@@ -63,7 +55,7 @@ public RegisteredService save(final RegisteredService registeredService) {
     @Override
     public synchronized RegisteredService save(final RegisteredService registeredService, final boolean publishEvent) {
         publishEvent(new CasRegisteredServicePreSaveEvent(this, registeredService));
-        val r = this.serviceRegistry.save(registeredService);
+        var r = this.serviceRegistry.save(registeredService);
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:58]: Is this change needed?

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 5]
```diff
@@ -104,10 +96,17 @@ public RegisteredService findServiceBy(final String serviceId) {
...
+                .filter(r -> r.matches(serviceId))
+                .findFirst()
+                .orElse(null);
+
+        if (service == null) {
+            service = serviceRegistry.findServiceBy(serviceId);
+            if (service != null) {
+                services.put(service.getId(), service);
+            }
+        }
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:109]: It would be good to add some log statements here under trace to explain the behavior and flow.

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 6]
```diff
@@ -118,22 +117,24 @@ public RegisteredService findServiceBy(final String serviceId) {
...
             return new ArrayList<>(0);
         }
-
-        return getAllServices()
-            .stream()
-            .filter(getRegisteredServicesFilteringPredicate(predicate))
-            .sorted()
-            .peek(RegisteredService::initialize)
-            .collect(Collectors.toList());
+        var results= serviceRegistry.findServicePredicate(predicate).
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:129]:
```suggestion
        var results = serviceRegistry.findServicePredicate(predicate).
```

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 7]
```diff
@@ -155,48 +156,122 @@ public RegisteredService findServiceBy(final Service service) {
 
     @Override
     public RegisteredService findServiceBy(final long id) {
-        val result = this.services.get(id);
+        var result = this.services.get(id, k -> this.serviceRegistry.findServiceById(id));
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:159]: Are you switching to var in order to correctly determine the final type?

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 8]
```diff
@@ -155,48 +156,122 @@ public RegisteredService findServiceBy(final Service service) {
...
+            service = serviceRegistry.findServiceByExactServiceName(name);
+            if (service != null) {
+                services.put(service.getId(), service);
+            }
+        }
+
+        if (service != null) {
+            service.initialize();
+        }
+        return validateRegisteredService(service);
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:191]: Similar to above, log statements under trace would be good.

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 9]
```diff
@@ -155,48 +156,122 @@ public RegisteredService findServiceBy(final Service service) {
...
+    }
+    
+    @Override
+    public <T extends RegisteredService> T findServiceByName(final String name, final Class<T> clazz) {
+        var result= this.serviceRegistry.findServiceByExactServiceName(name, clazz);
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:196]:
```suggestion
        var result = this.serviceRegistry.findServiceByExactServiceName(name, clazz);
```

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 10]
```diff
@@ -155,48 +156,122 @@ public RegisteredService findServiceBy(final Service service) {
...
+        services.get(result.getId(), k-> result);
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:197]: This won't work.

What happens if the result is null?

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 11]
```diff
@@ -155,48 +156,122 @@ public RegisteredService findServiceBy(final Service service) {
...
+            if (service != null) {
+                services.put(service.getId(), service);
+            }
+        }
+
+        if (service != null) {
+            service.initialize();
+        }
+        return validateRegisteredService(service);      
+    }
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:223]: Likewise, logs would be good.

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 12]
```diff
@@ -155,48 +156,122 @@ public RegisteredService findServiceBy(final Service service) {
...
-        this.services = this.serviceRegistry.load()
-            .stream()
-            .collect(Collectors.toConcurrentMap(r -> {
-                LOGGER.debug("Adding registered service [{}] with name [{}] and internal identifier [{}]", r.getServiceId(), r.getName(), r.getId());
-                return r.getId();
-            }, Function.identity(), (r, s) -> s));
+        this.services.putAll(this.serviceRegistry.load()
+                .stream()
+                .collect(Collectors.toMap(r -> {
+                    LOGGER.debug("Adding registered service [{}] with name [{}] and internal identifier [{}]",
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:247]:
```suggestion
                    LOGGER.trace("Adding registered service [{}] with name [{}] and internal identifier [{}]",
```

**Comment by mmoayyed** [comment]
Also, this is no big deal at all; but please make sure your tab/indentation size is set to 4. It seems like you have it set to 8. (No need to make fixes/corrections; what you have is fine.)

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 13]
```diff
@@ -104,10 +96,21 @@ public RegisteredService findServiceBy(final String serviceId) {
...
-            .filter(r -> r.matches(serviceId))
-            .findFirst()
-            .orElse(null);
+        var service = getCandidateServicesToMatch(serviceId)
+                .filter(r -> r.matches(serviceId))
+                .findFirst()
+                .orElse(null);
+
+        if (service == null) {
+            LOGGER.trace("The service that matches the serviceId {} is not found in the cache, try to find it from [{}]",
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:105]:
```suggestion
            LOGGER.trace("The service that matches the service id [{}] is not found in the cache; trying to find it from [{}]",
```

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 14]
```diff
@@ -104,10 +96,21 @@ public RegisteredService findServiceBy(final String serviceId) {
...
+                    serviceId, serviceRegistry.getName());
+            service = serviceRegistry.findServiceBy(serviceId);
+            if (service != null) {
+                services.put(service.getId(), service);
+                LOGGER.trace("The service is found in [{}] and populated to the cache {}  ", serviceRegistry.getName(),
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:110]:
```suggestion
                LOGGER.trace("The service is found in [{}] and populated to the cache [{}]", serviceRegistry.getName(), 
```

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 15]
```diff
@@ -155,48 +160,134 @@ public RegisteredService findServiceBy(final Service service) {
...
     }
 
+    @Override
+    public <T extends RegisteredService> T findServiceBy(final long id, final Class<T> clazz) {
+        val result = this.serviceRegistry.findServiceById(id, clazz);
+        if (result != null) {
+            services.get(result.getId(), k -> result);
+        }
+        return result;
+    }
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:174]: This doesn't look right. Should be:


```
result = services.get(result.getId(), k -> this.serviceRegistry.findServiceById(id, clazz));
```

Also, don't forget to check the result to make sure it's has the correct type based on `clazz`

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 16]
```diff
@@ -155,48 +160,134 @@ public RegisteredService findServiceBy(final Service service) {
...
+            return null;
+        }
+
+        var service = services.asMap().values().stream()
+                .filter(r -> r.getName().equals(name))
+                .findFirst()
+                .orElse(null);
+
+        if (service == null) {
+            LOGGER.trace("The service with name {} is not found in the cache, try to find it from [{}]",
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:188]:
```suggestion
            LOGGER.trace("The service with name [{}] is not found in the cache, try to find it from [{}]",
```

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 17]
```diff
@@ -155,48 +160,134 @@ public RegisteredService findServiceBy(final Service service) {
...
+                    name, serviceRegistry.getName());
+            service = serviceRegistry.findServiceByExactServiceName(name);
+            if (service != null) {
+                services.put(service.getId(), service);
+                LOGGER.trace("The service is found in [{}] and populated to the cache {}  ", serviceRegistry.getName(),
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:193]:
```suggestion
                LOGGER.trace("The service is found in [{}] and populated to the cache [{}]", serviceRegistry.getName(),
```

Also, formatting of parameters is off here. Should be: "The service [{}] is found in service registry [{}]" and cached.

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 18]
```diff
@@ -155,48 +160,134 @@ public RegisteredService findServiceBy(final Service service) {
...
+        return validateRegisteredService(service);
+    }
+    
+    @Override
+    public <T extends RegisteredService> T findServiceByName(final String name, final Class<T> clazz) {
+        val result = this.serviceRegistry.findServiceByExactServiceName(name, clazz);
+        if (result != null) {
+            services.get(result.getId(), k -> result);
+        }
+        return result;
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:210]: Same as above.

File: core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java [hunk 19]
```diff
@@ -155,48 +160,134 @@ public RegisteredService findServiceBy(final Service service) {
...
+                .orElse(null);
+
+        if (service == null) {
+            LOGGER.trace("The service with serviceId {} is not found in the cache, try to find it from [{}]",
+                    serviceId, serviceRegistry.getName());
+            service = serviceRegistry.findServiceByExactServiceId(serviceId);
+            if (service != null) {
+                services.put(service.getId(), service);
+                LOGGER.trace("The service is found in [{}] and populated to the cache {}  ", serviceRegistry.getName(),
+                        service);
```
Comment (mmoayyed) [inline core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java:231]: Same as above.

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/services/ServiceRegistryProperties.java [hunk 20]
```diff
@@ -164,7 +164,22 @@
      * </ul>
      */
     private ServiceManagementTypes managementType = ServiceManagementTypes.DEFAULT;
+    
+    /**
+     * services cache duration specifies the fixed duration for an entry to be automatically removed from the cache after its creation or update.
```
Comment (mmoayyed) [inline api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/services/ServiceRegistryProperties.java:169]:
```suggestion
     * Services cache duration specifies the fixed duration for an entry to be automatically removed from the cache after its creation or update.
```

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/services/ServiceRegistryProperties.java [hunk 21]
```diff
@@ -164,7 +164,22 @@
...
+    */
+    private String cache = "PT5M";
+
+    /**
+     * services cache size specifies the maximum number of entries the cache may contain.
```
Comment (mmoayyed) [inline api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/services/ServiceRegistryProperties.java:174]:
```suggestion
     * Services cache size specifies the maximum number of entries the cache may contain.
```

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/services/ServiceRegistryProperties.java [hunk 22]
```diff
@@ -164,7 +164,22 @@
...
+    */
+    private long cacheSize = Long.MAX_VALUE;
 
+    /**
+     * services cache capacity sets the minimum total size for the internal data structures.
```
Comment (mmoayyed) [inline api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/services/ServiceRegistryProperties.java:179]:
```suggestion
     * Services cache capacity sets the minimum total size for the internal data structures.
```

**Comment by mmoayyed** [comment]
Also, forgot to mention that all new properties should be documented in manual too.
